model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/organization/InfinispanOrganizationProvider.java

```diff
@@ -0,0 +1,222 @@
+        }
+
+        if (cached == null) {
+            Long loaded = realmCache.getCache().getCurrentRevision(cacheKey);
+            OrganizationModel model = jpaOrgDelegate.getByDomainName(domainName);
+            if (model == null) return null;
+            if (realmCache.getInvalidations().contains(model.getId())) return model;
+            cached = new CachedOrganization(loaded, getRealm(), model);
+            realmCache.getCache().addRevisioned(cached, realmCache.getStartupRevision());
+
+        // no need to check for realm invalidation as IdP changes are handled by events within InfinispanOrganizationProviderFactory
+        } else if (realmCache.getInvalidations().contains(cached.getId())) {
+            return getCacheDelegate(jpaOrgDelegate.getByDomainName(domainName));
+        } else if (managedOrganizations.containsKey(cached.getId())) {
+            return managedOrganizations.get(cached.getId());
+        }
+        OrganizationAdapter adapter = new OrganizationAdapter(cached, realmCache, jpaOrgDelegate);
+        managedOrganizations.put(cacheKey, adapter);
+        return adapter;
+    }
+
+    @Override
+    public Stream<OrganizationModel> getAllStream(String search, Boolean exact, Integer first, Integer max) {
+        // Return cache delegates to ensure cache invalidation during write operations
+        return getCacheDelegates(jpaOrgDelegate.getAllStream(search, exact, first, max));
+    }
+
+    @Override
+    public Stream<OrganizationModel> getAllStream(Map<String, String> attributes, Integer first, Integer max) {
+        // Return cache delegates to ensure cache invalidation during write operations
+        return getCacheDelegates(jpaOrgDelegate.getAllStream(attributes, first, max));
+    }
+
+    @Override
+    public boolean remove(OrganizationModel organization) {
+        registerOrganizationInvalidation(organization.getId());
+        return jpaOrgDelegate.remove(organization);
+    }
+
+    @Override
+    public void removeAll() {
+        //TODO: won't scale, requires a better mechanism for bulk deleting organizations within a realm
+        //this way, all organizations in the realm will be invalidated ... or should it be invalidated whole realm instead?
+        getAllStream().forEach(this::remove);
+    }
+
+    @Override
+    public boolean addMember(OrganizationModel organization, UserModel user) {
+        return jpaOrgDelegate.addMember(organization, user);
+    }
+
+    @Override
+    public boolean removeMember(OrganizationModel organization, UserModel member) {
+        return jpaOrgDelegate.removeMember(organization, member);
+    }
+
+    @Override
+    public Stream<UserModel> getMembersStream(OrganizationModel organization, String search, Boolean exact, Integer first, Integer max) {
+        return jpaOrgDelegate.getMembersStream(organization, search, exact, first, max);
+    }
+
+    @Override
+    public UserModel getMemberById(OrganizationModel organization, String id) {
```

Comment:
This method is used in called from critical places like during authentication and token mapping. Can we have a cache here too?

Reply:
To make it work then changes made to this user need to invalidate the org cache too.. perhaps add a few more listeners to the factory so they invalidate the user cache in this case?

Reply:
I'm not sure I understand what do we want to cache here. Looking at the implementation it 
- gets user by the provided `id` from user provider (either from cache or from db)
- if a user is found, obtain the `orgId` from its attributes and checks whether it matches with the provided `organization`, if so return the user, otherwise return `null`.

Reply:
True, nothing to cache here. I missed what we are doing already ...